custom_field/import_custom_fields.rb

```diff
@@ -0,0 +1,284 @@
+  #
+  # @param api_token [String] The PagerDuty API token.
+  # @param logger [Logger] The logger instance for logging information and errors.
+  def initialize(api_token, logger)
+    @api_token = api_token
+    @logger = logger
+    @existing_fields = fetch_existing_custom_fields
+  end
+
+  # Sends an API request to PagerDuty to create a custom field.
+  #
+  # @param params [Hash] A hash containing the custom field parameters.
+  # @return [Hash, nil] A hash containing the response from the API, or nil if the request was unsuccessful.
+  # @see {PagerDuty API Documentation}[https://developer.pagerduty.com/api-reference/0f6094f852517-update-custom-field-values]
+  def create_custom_field(params)
+    endpoint = '/incidents/custom_fields'
+    options = {
+      body: { 'field': params }.to_json,
+      headers: { 'Authorization' => "Token token=#{@api_token}" }
+    }
+
+    @logger.debug 'Calling private function to determine if custom field exists'
+    if check_existing_custom_field(params[:name])
+      @logger.info "Skipping attempt to create #{params[:name]}: already exists."
+      puts "Skipping creation of #{params[:name]}"
+      puts "     a custom field by that name alread exists on this api-key's account."
+      return { 'status' => 'skipped' }
+    end
+
+    @logger.info "Sending POST request to #{endpoint} with body: #{options[:body]}"
+    response = self.class.post(endpoint, options)
+    parsed_response = JSON.parse(response.body)
+
+    if response.success?
+      @logger.info "Successfully created custom field. Response: #{parsed_response}"
+      parsed_response
+    else
+      handle_api_error(parsed_response)
+    end
+  rescue HTTParty::Error => e
+    handle_connection_error(e)
+  end
+
+  private
+
+  # Fetches existing custom fields from PagerDuty.
+  #
+  # @return [Array, false] List of existing fields or false if request failed.
+  def fetch_existing_custom_fields
+    endpoint = '/incidents/custom_fields'
+    options = {
+      headers: {
+        'Authorization' => "Token token=#{@api_token}",
+        'Content-Type' => 'application/json'
+      }
+    }
+    @logger.info "Making call to get existing field names from #{endpoint}"
+    response = self.class.get(endpoint, options)
+    @logger.info "The following response was received: #{response}"
+    if response.success?
+      json = JSON.parse(response.body)
+      @logger.info "The following parsing is being stored: #{json}"
+      json
+    else
+      handle_api_error(JSON.parse(response.body))
+    end
+  rescue HTTParty::Error => e
+    handle_connection_error(e)
+  end
+
+  # Checks if a custom field with the provided name already exists.
+  #
+  # @param field_name [String] The name of the custom field to check for.
+  # @return [Boolean] Returns true if the custom field with the given name exists,
+  # otherwise returns false (including if it is unabel to determine the result).
+  def check_existing_custom_field(field_name)
```

Comment:
@ethanmsl the PagerDuty API validates the uniqueness of the name, I'm not sure we need to burden the script with this check when the API already takes care of it, unless I am missing something.

Reply:
Two elements here: result & structure:

Result-wise: 
The script returns an error if the response is not a success.  (including a 2001 error from PD for a name collision) 
In general if someone asks for a list of names (custom fields) to be on an account what they care about is that the names are there not that they added them.  So adding the ability to fail gracefully on existing  means that the script can acheive its result regardless of current state, while still informing the customer.

As a practical point: people are often going to want to keep a single goal list and re-use it.  So we want the script to continue on *this* error.  (But fail out on most other errors.)

Structure-wise:
We could have just implemented this as error handling on the PUT call.  But having our defacto "GET" logic be a side effect of a PUT -- means that many changes to this script would require re-working error handling and re-implementing the GET.  So it's just simpler to throw a GET in there.

If anyone wants to report on custom fields, selectively delete, etc.  it will be there.  And we already needed the unerlying logic in order to achieve the script's original aim. 

Filewise -- if the script were expanded in the future I'd support breaking it into multiple files.  But right now it's just a GET, PUT, and PARSE mostly -- with the GET & PUT being almost identical.

Reply:
That sounds good 👍  - I think we should merge and keep this PR no longer than what it is - any new iteration should be in the form of a new PR - awesome work!

Reply:
I have tested the new changes and I think it's good to 🚀